# 03 — Train Orthoptera Classifier
Fine-tune a CNN on the prepared label splits using OpenSoundscape, then export the trained model to `models/` for use in BASE.

**Kernel:** `Python (orthoptera-training)`  
**Prerequisites:** Run `02_prepare_labels.ipynb` first to generate `training/data/train.csv` etc.

In [1]:
import torch
import pandas as pd
from pathlib import Path
import opensoundscape.ml.cnn
from opensoundscape import CNN

PROJECT_ROOT = Path.cwd().parent.parent
DATA_DIR   = PROJECT_ROOT / "training" / "data"
MODELS_DIR = PROJECT_ROOT / "models"
MODELS_DIR.mkdir(exist_ok=True)

train = pd.read_csv(DATA_DIR / "train.csv", index_col=[0, 1, 2])
val   = pd.read_csv(DATA_DIR / "val.csv",   index_col=[0, 1, 2])
test  = pd.read_csv(DATA_DIR / "test.csv",  index_col=[0, 1, 2])

print(f"Classes: {list(train.columns)}")
print(f"Train: {len(train)}  Val: {len(val)}  Test: {len(test)}")

Classes: ['Chorthippus brunneus brunneus', 'Pseudochorthippus parallelus parallelus', 'Omocestus viridulus', 'Tettigonia viridissima', 'Roeseliana roeselii', 'Pholidoptera griseoaptera', 'Leptophyes punctatissima', 'Gryllus campestris']
Train: 18314  Val: 6436  Test: 4631


In [2]:
# ── Build model ──────────────────────────────────────────────────────────────
model = CNN(
    architecture="resnet18",
    classes=list(train.columns),
    sample_duration=4.0,
)
print("Model built — architecture: resnet18")
print(f"Classes: {model.classes}")
model.network.to("mps")
print(next(model.network.parameters()).device)


Model built — architecture: resnet18
Classes: ['Chorthippus brunneus brunneus', 'Pseudochorthippus parallelus parallelus', 'Omocestus viridulus', 'Tettigonia viridissima', 'Roeseliana roeselii', 'Pholidoptera griseoaptera', 'Leptophyes punctatissima', 'Gryllus campestris']
mps:0


In [3]:
print(opensoundscape.ml.cnn_architectures.list_architectures())

['resnet18', 'resnet34', 'resnet50', 'resnet101', 'resnet152', 'alexnet', 'vgg11_bn', 'squeezenet1_0', 'densenet121', 'inception_v3', 'efficientnet_b0', 'efficientnet_b4', 'efficientnet_widese_b0', 'efficientnet_widese_b4']


In [4]:
# ── Train ────────────────────────────────────────────────────────────────────
# Run config: 5 epochs, batch_size=32, num_workers=2, CPU.
# ~7-8h per epoch on CPU (~40h total). Use a GPU machine to cut this to minutes.

model.train(
    train,
    val,
    epochs=5,
    batch_size=32,
    save_path=MODELS_DIR / "orthoptera_checkpoints",
    save_interval=1,
    num_workers=2,
)



Training Epoch 0


  0%|          | 0/573 [00:00<?, ?it/s]

/Users/dawsonmccall/Documents/coding_projects/BioAcousticStreamEngine/.venv-training/lib/python3.11/site-packages/torch/utils/data/dataloader.py:1095: UserWarning: 'pin_memory' argument is set as true but not supported on MPS now, device pinned memory won't be used.
  super().__init__(loader)


Epoch: 0 [batch 0/573, 0.00%] 
	Epoch Running Average Loss: 0.837
	Most Recent Batch Loss: 0.837
Epoch: 0 [batch 10/573, 1.75%] 
	Epoch Running Average Loss: 0.438
	Most Recent Batch Loss: 0.322
Epoch: 0 [batch 20/573, 3.49%] 
	Epoch Running Average Loss: 0.373
	Most Recent Batch Loss: 0.237
Epoch: 0 [batch 30/573, 5.24%] 
	Epoch Running Average Loss: 0.339
	Most Recent Batch Loss: 0.285


[src/libmpg123/layer3.c:INT123_do_layer3():1776] error: part2_3_length (3744) too large for available bit count (3736)


Epoch: 0 [batch 40/573, 6.98%] 
	Epoch Running Average Loss: 0.318
	Most Recent Batch Loss: 0.252
Epoch: 0 [batch 50/573, 8.73%] 
	Epoch Running Average Loss: 0.303
	Most Recent Batch Loss: 0.227


/Users/dawsonmccall/Documents/coding_projects/BioAcousticStreamEngine/.venv-training/lib/python3.11/site-packages/opensoundscape/audio.py:1411: UserWarning: This seems to be an AudioMoth file, but parse_audiomoth_metadata() raised: 'comment'
  warnings.warn(


Epoch: 0 [batch 60/573, 10.47%] 
	Epoch Running Average Loss: 0.293
	Most Recent Batch Loss: 0.240
Epoch: 0 [batch 70/573, 12.22%] 
	Epoch Running Average Loss: 0.281
	Most Recent Batch Loss: 0.190
Epoch: 0 [batch 80/573, 13.96%] 
	Epoch Running Average Loss: 0.273
	Most Recent Batch Loss: 0.229
Epoch: 0 [batch 90/573, 15.71%] 
	Epoch Running Average Loss: 0.265
	Most Recent Batch Loss: 0.192
Epoch: 0 [batch 100/573, 17.45%] 
	Epoch Running Average Loss: 0.260
	Most Recent Batch Loss: 0.211
Epoch: 0 [batch 110/573, 19.20%] 
	Epoch Running Average Loss: 0.254
	Most Recent Batch Loss: 0.205
Epoch: 0 [batch 120/573, 20.94%] 
	Epoch Running Average Loss: 0.250
	Most Recent Batch Loss: 0.226
Epoch: 0 [batch 130/573, 22.69%] 
	Epoch Running Average Loss: 0.245
	Most Recent Batch Loss: 0.178
Epoch: 0 [batch 140/573, 24.43%] 
	Epoch Running Average Loss: 0.241
	Most Recent Batch Loss: 0.158
Epoch: 0 [batch 150/573, 26.18%] 
	Epoch Running Average Loss: 0.237
	Most Recent Batch Loss: 0.223


/Users/dawsonmccall/Documents/coding_projects/BioAcousticStreamEngine/.venv-training/lib/python3.11/site-packages/opensoundscape/audio.py:1411: UserWarning: This seems to be an AudioMoth file, but parse_audiomoth_metadata() raised: time data '01:00:00 14/08/202)' does not match format '%H:%M:%S %d/%m/%Y (%Z%z)'
  warnings.warn(


Epoch: 0 [batch 160/573, 27.92%] 
	Epoch Running Average Loss: 0.233
	Most Recent Batch Loss: 0.223
Epoch: 0 [batch 170/573, 29.67%] 
	Epoch Running Average Loss: 0.230
	Most Recent Batch Loss: 0.167
Epoch: 0 [batch 180/573, 31.41%] 
	Epoch Running Average Loss: 0.228
	Most Recent Batch Loss: 0.184
Epoch: 0 [batch 190/573, 33.16%] 
	Epoch Running Average Loss: 0.225
	Most Recent Batch Loss: 0.130
Epoch: 0 [batch 200/573, 34.90%] 
	Epoch Running Average Loss: 0.222
	Most Recent Batch Loss: 0.196
Epoch: 0 [batch 210/573, 36.65%] 
	Epoch Running Average Loss: 0.220
	Most Recent Batch Loss: 0.158
Epoch: 0 [batch 220/573, 38.39%] 
	Epoch Running Average Loss: 0.217
	Most Recent Batch Loss: 0.184
Epoch: 0 [batch 230/573, 40.14%] 
	Epoch Running Average Loss: 0.215
	Most Recent Batch Loss: 0.147
Epoch: 0 [batch 240/573, 41.88%] 
	Epoch Running Average Loss: 0.213
	Most Recent Batch Loss: 0.171
Epoch: 0 [batch 250/573, 43.63%] 
	Epoch Running Average Loss: 0.210
	Most Recent Batch Loss: 0.127


[src/libmpg123/layer3.c:INT123_do_layer3():1776] error: part2_3_length (1632) too large for available bit count (1560)


Epoch: 0 [batch 350/573, 61.08%] 
	Epoch Running Average Loss: 0.194
	Most Recent Batch Loss: 0.111
Epoch: 0 [batch 360/573, 62.83%] 
	Epoch Running Average Loss: 0.193
	Most Recent Batch Loss: 0.116
Epoch: 0 [batch 370/573, 64.57%] 
	Epoch Running Average Loss: 0.192
	Most Recent Batch Loss: 0.111
Epoch: 0 [batch 380/573, 66.32%] 
	Epoch Running Average Loss: 0.190
	Most Recent Batch Loss: 0.194
Epoch: 0 [batch 390/573, 68.06%] 
	Epoch Running Average Loss: 0.189
	Most Recent Batch Loss: 0.153
Epoch: 0 [batch 400/573, 69.81%] 
	Epoch Running Average Loss: 0.188
	Most Recent Batch Loss: 0.169
Epoch: 0 [batch 410/573, 71.55%] 
	Epoch Running Average Loss: 0.187
	Most Recent Batch Loss: 0.146
Epoch: 0 [batch 420/573, 73.30%] 
	Epoch Running Average Loss: 0.185
	Most Recent Batch Loss: 0.113
Epoch: 0 [batch 430/573, 75.04%] 
	Epoch Running Average Loss: 0.184
	Most Recent Batch Loss: 0.170
Epoch: 0 [batch 440/573, 76.79%] 
	Epoch Running Average Loss: 0.183
	Most Recent Batch Loss: 0.129


[src/libmpg123/layer3.c:INT123_do_layer3():1776] error: part2_3_length (1920) too large for available bit count (1912)


Epoch: 0 [batch 450/573, 78.53%] 
	Epoch Running Average Loss: 0.183
	Most Recent Batch Loss: 0.157
Epoch: 0 [batch 460/573, 80.28%] 
	Epoch Running Average Loss: 0.182
	Most Recent Batch Loss: 0.126
Epoch: 0 [batch 470/573, 82.02%] 
	Epoch Running Average Loss: 0.181
	Most Recent Batch Loss: 0.119
Epoch: 0 [batch 480/573, 83.77%] 
	Epoch Running Average Loss: 0.180
	Most Recent Batch Loss: 0.122
Epoch: 0 [batch 490/573, 85.51%] 
	Epoch Running Average Loss: 0.179
	Most Recent Batch Loss: 0.096
Epoch: 0 [batch 500/573, 87.26%] 
	Epoch Running Average Loss: 0.177
	Most Recent Batch Loss: 0.134
Epoch: 0 [batch 510/573, 89.01%] 
	Epoch Running Average Loss: 0.177
	Most Recent Batch Loss: 0.173
Epoch: 0 [batch 520/573, 90.75%] 
	Epoch Running Average Loss: 0.175
	Most Recent Batch Loss: 0.100
Epoch: 0 [batch 530/573, 92.50%] 
	Epoch Running Average Loss: 0.174
	Most Recent Batch Loss: 0.166
Epoch: 0 [batch 540/573, 94.24%] 
	Epoch Running Average Loss: 0.173
	Most Recent Batch Loss: 0.120


/Users/dawsonmccall/.pyenv/versions/3.11.9/lib/python3.11/tempfile.py:934: ResourceWarning: Implicitly cleaning up <TemporaryDirectory '/var/folders/dc/5p9wmvyd0m3c1sylx0z9lwlw0000gn/T/tmp4brtdrxpwandb'>
  _warnings.warn(warn_message, ResourceWarning)
/Users/dawsonmccall/.pyenv/versions/3.11.9/lib/python3.11/tempfile.py:934: ResourceWarning: Implicitly cleaning up <TemporaryDirectory '/var/folders/dc/5p9wmvyd0m3c1sylx0z9lwlw0000gn/T/tmp1g93skerwandb-artifacts'>
  _warnings.warn(warn_message, ResourceWarning)
/Users/dawsonmccall/.pyenv/versions/3.11.9/lib/python3.11/tempfile.py:934: ResourceWarning: Implicitly cleaning up <TemporaryDirectory '/var/folders/dc/5p9wmvyd0m3c1sylx0z9lwlw0000gn/T/tmp2pydjtj4wandb-media'>
  _warnings.warn(warn_message, ResourceWarning)
/Users/dawsonmccall/.pyenv/versions/3.11.9/lib/python3.11/tempfile.py:934: ResourceWarning: Implicitly cleaning up <TemporaryDirectory '/var/folders/dc/5p9wmvyd0m3c1sylx0z9lwlw0000gn/T/tmpoiqf9kxmwandb-media'>
  _warnings.warn(w

Metrics:
	MAP: 0.409

Validation.


  0%|          | 0/202 [00:00<?, ?it/s]

/Users/dawsonmccall/.pyenv/versions/3.11.9/lib/python3.11/tempfile.py:934: ResourceWarning: Implicitly cleaning up <TemporaryDirectory '/var/folders/dc/5p9wmvyd0m3c1sylx0z9lwlw0000gn/T/tmpde4_n1l0wandb'>
  _warnings.warn(warn_message, ResourceWarning)
/Users/dawsonmccall/.pyenv/versions/3.11.9/lib/python3.11/tempfile.py:934: ResourceWarning: Implicitly cleaning up <TemporaryDirectory '/var/folders/dc/5p9wmvyd0m3c1sylx0z9lwlw0000gn/T/tmp50tqb6bgwandb'>
  _warnings.warn(warn_message, ResourceWarning)
/Users/dawsonmccall/.pyenv/versions/3.11.9/lib/python3.11/tempfile.py:934: ResourceWarning: Implicitly cleaning up <TemporaryDirectory '/var/folders/dc/5p9wmvyd0m3c1sylx0z9lwlw0000gn/T/tmp3upmkwz8wandb-artifacts'>
  _warnings.warn(warn_message, ResourceWarning)
/Users/dawsonmccall/.pyenv/versions/3.11.9/lib/python3.11/tempfile.py:934: ResourceWarning: Implicitly cleaning up <TemporaryDirectory '/var/folders/dc/5p9wmvyd0m3c1sylx0z9lwlw0000gn/T/tmpu5u_95wuwandb-artifacts'>
  _warnings.warn(war

Metrics:
	MAP: 0.441

Training Epoch 1


  0%|          | 0/573 [00:00<?, ?it/s]

/Users/dawsonmccall/Documents/coding_projects/BioAcousticStreamEngine/.venv-training/lib/python3.11/site-packages/torch/utils/data/dataloader.py:1095: UserWarning: 'pin_memory' argument is set as true but not supported on MPS now, device pinned memory won't be used.
  super().__init__(loader)


Epoch: 1 [batch 0/573, 0.00%] 
	Epoch Running Average Loss: 0.133
	Most Recent Batch Loss: 0.133
Epoch: 1 [batch 10/573, 1.75%] 
	Epoch Running Average Loss: 0.120
	Most Recent Batch Loss: 0.130
Epoch: 1 [batch 20/573, 3.49%] 
	Epoch Running Average Loss: 0.120
	Most Recent Batch Loss: 0.147
Epoch: 1 [batch 30/573, 5.24%] 
	Epoch Running Average Loss: 0.120
	Most Recent Batch Loss: 0.195
Epoch: 1 [batch 40/573, 6.98%] 
	Epoch Running Average Loss: 0.120
	Most Recent Batch Loss: 0.080
Epoch: 1 [batch 50/573, 8.73%] 
	Epoch Running Average Loss: 0.122
	Most Recent Batch Loss: 0.086
Epoch: 1 [batch 60/573, 10.47%] 
	Epoch Running Average Loss: 0.120
	Most Recent Batch Loss: 0.088
Epoch: 1 [batch 70/573, 12.22%] 
	Epoch Running Average Loss: 0.120
	Most Recent Batch Loss: 0.087
Epoch: 1 [batch 80/573, 13.96%] 
	Epoch Running Average Loss: 0.119
	Most Recent Batch Loss: 0.147
Epoch: 1 [batch 90/573, 15.71%] 
	Epoch Running Average Loss: 0.118
	Most Recent Batch Loss: 0.089
Epoch: 1 [batch 1

/Users/dawsonmccall/Documents/coding_projects/BioAcousticStreamEngine/.venv-training/lib/python3.11/site-packages/opensoundscape/audio.py:1411: UserWarning: This seems to be an AudioMoth file, but parse_audiomoth_metadata() raised: time data '01:00:00 14/08/202)' does not match format '%H:%M:%S %d/%m/%Y (%Z%z)'
  warnings.warn(


Epoch: 1 [batch 110/573, 19.20%] 
	Epoch Running Average Loss: 0.119
	Most Recent Batch Loss: 0.095
Epoch: 1 [batch 120/573, 20.94%] 
	Epoch Running Average Loss: 0.119
	Most Recent Batch Loss: 0.128
Epoch: 1 [batch 130/573, 22.69%] 
	Epoch Running Average Loss: 0.119
	Most Recent Batch Loss: 0.096
Epoch: 1 [batch 140/573, 24.43%] 
	Epoch Running Average Loss: 0.119
	Most Recent Batch Loss: 0.101
Epoch: 1 [batch 150/573, 26.18%] 
	Epoch Running Average Loss: 0.120
	Most Recent Batch Loss: 0.121
Epoch: 1 [batch 160/573, 27.92%] 
	Epoch Running Average Loss: 0.119
	Most Recent Batch Loss: 0.113
Epoch: 1 [batch 170/573, 29.67%] 
	Epoch Running Average Loss: 0.119
	Most Recent Batch Loss: 0.102
Epoch: 1 [batch 180/573, 31.41%] 
	Epoch Running Average Loss: 0.118
	Most Recent Batch Loss: 0.094
Epoch: 1 [batch 190/573, 33.16%] 
	Epoch Running Average Loss: 0.117
	Most Recent Batch Loss: 0.106
Epoch: 1 [batch 200/573, 34.90%] 
	Epoch Running Average Loss: 0.117
	Most Recent Batch Loss: 0.088


[src/libmpg123/layer3.c:INT123_do_layer3():1776] error: part2_3_length (1920) too large for available bit count (1912)


Epoch: 1 [batch 290/573, 50.61%] 
	Epoch Running Average Loss: 0.115
	Most Recent Batch Loss: 0.072
Epoch: 1 [batch 300/573, 52.36%] 
	Epoch Running Average Loss: 0.115
	Most Recent Batch Loss: 0.123
Epoch: 1 [batch 310/573, 54.10%] 
	Epoch Running Average Loss: 0.115
	Most Recent Batch Loss: 0.127
Epoch: 1 [batch 320/573, 55.85%] 
	Epoch Running Average Loss: 0.114
	Most Recent Batch Loss: 0.086
Epoch: 1 [batch 330/573, 57.59%] 
	Epoch Running Average Loss: 0.114
	Most Recent Batch Loss: 0.139
Epoch: 1 [batch 340/573, 59.34%] 
	Epoch Running Average Loss: 0.114
	Most Recent Batch Loss: 0.131
Epoch: 1 [batch 350/573, 61.08%] 
	Epoch Running Average Loss: 0.114
	Most Recent Batch Loss: 0.124
Epoch: 1 [batch 360/573, 62.83%] 
	Epoch Running Average Loss: 0.114
	Most Recent Batch Loss: 0.117


[src/libmpg123/layer3.c:INT123_do_layer3():1776] error: part2_3_length (1632) too large for available bit count (1560)


Epoch: 1 [batch 370/573, 64.57%] 
	Epoch Running Average Loss: 0.113
	Most Recent Batch Loss: 0.132
Epoch: 1 [batch 380/573, 66.32%] 
	Epoch Running Average Loss: 0.114
	Most Recent Batch Loss: 0.080
Epoch: 1 [batch 390/573, 68.06%] 
	Epoch Running Average Loss: 0.114
	Most Recent Batch Loss: 0.078


/Users/dawsonmccall/Documents/coding_projects/BioAcousticStreamEngine/.venv-training/lib/python3.11/site-packages/opensoundscape/audio.py:1411: UserWarning: This seems to be an AudioMoth file, but parse_audiomoth_metadata() raised: 'comment'
  warnings.warn(


Epoch: 1 [batch 400/573, 69.81%] 
	Epoch Running Average Loss: 0.113
	Most Recent Batch Loss: 0.114
Epoch: 1 [batch 410/573, 71.55%] 
	Epoch Running Average Loss: 0.113
	Most Recent Batch Loss: 0.090
Epoch: 1 [batch 420/573, 73.30%] 
	Epoch Running Average Loss: 0.113
	Most Recent Batch Loss: 0.089
Epoch: 1 [batch 430/573, 75.04%] 
	Epoch Running Average Loss: 0.113
	Most Recent Batch Loss: 0.126
Epoch: 1 [batch 440/573, 76.79%] 
	Epoch Running Average Loss: 0.113
	Most Recent Batch Loss: 0.087
Epoch: 1 [batch 450/573, 78.53%] 
	Epoch Running Average Loss: 0.113
	Most Recent Batch Loss: 0.084
Epoch: 1 [batch 460/573, 80.28%] 
	Epoch Running Average Loss: 0.112
	Most Recent Batch Loss: 0.093
Epoch: 1 [batch 470/573, 82.02%] 
	Epoch Running Average Loss: 0.112
	Most Recent Batch Loss: 0.104
Epoch: 1 [batch 480/573, 83.77%] 
	Epoch Running Average Loss: 0.112
	Most Recent Batch Loss: 0.087
Epoch: 1 [batch 490/573, 85.51%] 
	Epoch Running Average Loss: 0.112
	Most Recent Batch Loss: 0.097


[src/libmpg123/layer3.c:INT123_do_layer3():1776] error: part2_3_length (3744) too large for available bit count (3736)


Epoch: 1 [batch 510/573, 89.01%] 
	Epoch Running Average Loss: 0.112
	Most Recent Batch Loss: 0.116
Epoch: 1 [batch 520/573, 90.75%] 
	Epoch Running Average Loss: 0.111
	Most Recent Batch Loss: 0.139
Epoch: 1 [batch 530/573, 92.50%] 
	Epoch Running Average Loss: 0.111
	Most Recent Batch Loss: 0.108
Epoch: 1 [batch 540/573, 94.24%] 
	Epoch Running Average Loss: 0.111
	Most Recent Batch Loss: 0.076
Epoch: 1 [batch 550/573, 95.99%] 
	Epoch Running Average Loss: 0.111
	Most Recent Batch Loss: 0.067
Epoch: 1 [batch 560/573, 97.73%] 
	Epoch Running Average Loss: 0.110
	Most Recent Batch Loss: 0.126
Epoch: 1 [batch 570/573, 99.48%] 
	Epoch Running Average Loss: 0.110
	Most Recent Batch Loss: 0.055


/Users/dawsonmccall/.pyenv/versions/3.11.9/lib/python3.11/tempfile.py:934: ResourceWarning: Implicitly cleaning up <TemporaryDirectory '/var/folders/dc/5p9wmvyd0m3c1sylx0z9lwlw0000gn/T/tmpssnzgi_bwandb'>
  _warnings.warn(warn_message, ResourceWarning)
/Users/dawsonmccall/.pyenv/versions/3.11.9/lib/python3.11/tempfile.py:934: ResourceWarning: Implicitly cleaning up <TemporaryDirectory '/var/folders/dc/5p9wmvyd0m3c1sylx0z9lwlw0000gn/T/tmpup742rrswandb-artifacts'>
  _warnings.warn(warn_message, ResourceWarning)
/Users/dawsonmccall/.pyenv/versions/3.11.9/lib/python3.11/tempfile.py:934: ResourceWarning: Implicitly cleaning up <TemporaryDirectory '/var/folders/dc/5p9wmvyd0m3c1sylx0z9lwlw0000gn/T/tmp9t26972wwandb-media'>
  _warnings.warn(warn_message, ResourceWarning)
/Users/dawsonmccall/.pyenv/versions/3.11.9/lib/python3.11/tempfile.py:934: ResourceWarning: Implicitly cleaning up <TemporaryDirectory '/var/folders/dc/5p9wmvyd0m3c1sylx0z9lwlw0000gn/T/tmp2z1cxvvcwandb'>
  _warnings.warn(warn_me

Metrics:
	MAP: 0.670

Validation.


  0%|          | 0/202 [00:00<?, ?it/s]

/Users/dawsonmccall/.pyenv/versions/3.11.9/lib/python3.11/tempfile.py:934: ResourceWarning: Implicitly cleaning up <TemporaryDirectory '/var/folders/dc/5p9wmvyd0m3c1sylx0z9lwlw0000gn/T/tmpuyzx_plowandb'>
  _warnings.warn(warn_message, ResourceWarning)
/Users/dawsonmccall/.pyenv/versions/3.11.9/lib/python3.11/tempfile.py:934: ResourceWarning: Implicitly cleaning up <TemporaryDirectory '/var/folders/dc/5p9wmvyd0m3c1sylx0z9lwlw0000gn/T/tmp1g_w7ecrwandb-artifacts'>
  _warnings.warn(warn_message, ResourceWarning)
/Users/dawsonmccall/.pyenv/versions/3.11.9/lib/python3.11/tempfile.py:934: ResourceWarning: Implicitly cleaning up <TemporaryDirectory '/var/folders/dc/5p9wmvyd0m3c1sylx0z9lwlw0000gn/T/tmpm0yunwy3wandb-media'>
  _warnings.warn(warn_message, ResourceWarning)
/Users/dawsonmccall/.pyenv/versions/3.11.9/lib/python3.11/tempfile.py:934: ResourceWarning: Implicitly cleaning up <TemporaryDirectory '/var/folders/dc/5p9wmvyd0m3c1sylx0z9lwlw0000gn/T/tmpzmxk4up1wandb-media'>
  _warnings.warn(w

Metrics:
	MAP: 0.540

Training Epoch 2


  0%|          | 0/573 [00:00<?, ?it/s]

/Users/dawsonmccall/Documents/coding_projects/BioAcousticStreamEngine/.venv-training/lib/python3.11/site-packages/torch/utils/data/dataloader.py:1095: UserWarning: 'pin_memory' argument is set as true but not supported on MPS now, device pinned memory won't be used.
  super().__init__(loader)


Epoch: 2 [batch 0/573, 0.00%] 
	Epoch Running Average Loss: 0.104
	Most Recent Batch Loss: 0.104
Epoch: 2 [batch 10/573, 1.75%] 
	Epoch Running Average Loss: 0.104
	Most Recent Batch Loss: 0.094
Epoch: 2 [batch 20/573, 3.49%] 
	Epoch Running Average Loss: 0.101
	Most Recent Batch Loss: 0.104
Epoch: 2 [batch 30/573, 5.24%] 
	Epoch Running Average Loss: 0.098
	Most Recent Batch Loss: 0.086
Epoch: 2 [batch 40/573, 6.98%] 
	Epoch Running Average Loss: 0.098
	Most Recent Batch Loss: 0.110
Epoch: 2 [batch 50/573, 8.73%] 
	Epoch Running Average Loss: 0.098
	Most Recent Batch Loss: 0.128
Epoch: 2 [batch 60/573, 10.47%] 
	Epoch Running Average Loss: 0.097
	Most Recent Batch Loss: 0.094
Epoch: 2 [batch 70/573, 12.22%] 
	Epoch Running Average Loss: 0.098
	Most Recent Batch Loss: 0.102
Epoch: 2 [batch 80/573, 13.96%] 
	Epoch Running Average Loss: 0.098
	Most Recent Batch Loss: 0.127
Epoch: 2 [batch 90/573, 15.71%] 
	Epoch Running Average Loss: 0.098
	Most Recent Batch Loss: 0.102
Epoch: 2 [batch 1

[src/libmpg123/layer3.c:INT123_do_layer3():1776] error: part2_3_length (1920) too large for available bit count (1912)


Epoch: 2 [batch 120/573, 20.94%] 
	Epoch Running Average Loss: 0.097
	Most Recent Batch Loss: 0.077
Epoch: 2 [batch 130/573, 22.69%] 
	Epoch Running Average Loss: 0.096
	Most Recent Batch Loss: 0.059
Epoch: 2 [batch 140/573, 24.43%] 
	Epoch Running Average Loss: 0.096
	Most Recent Batch Loss: 0.141
Epoch: 2 [batch 150/573, 26.18%] 
	Epoch Running Average Loss: 0.097
	Most Recent Batch Loss: 0.102
Epoch: 2 [batch 160/573, 27.92%] 
	Epoch Running Average Loss: 0.097
	Most Recent Batch Loss: 0.087
Epoch: 2 [batch 170/573, 29.67%] 
	Epoch Running Average Loss: 0.097
	Most Recent Batch Loss: 0.106
Epoch: 2 [batch 180/573, 31.41%] 
	Epoch Running Average Loss: 0.098
	Most Recent Batch Loss: 0.084
Epoch: 2 [batch 190/573, 33.16%] 
	Epoch Running Average Loss: 0.098
	Most Recent Batch Loss: 0.074
Epoch: 2 [batch 200/573, 34.90%] 
	Epoch Running Average Loss: 0.097
	Most Recent Batch Loss: 0.112
Epoch: 2 [batch 210/573, 36.65%] 
	Epoch Running Average Loss: 0.097
	Most Recent Batch Loss: 0.093


/Users/dawsonmccall/Documents/coding_projects/BioAcousticStreamEngine/.venv-training/lib/python3.11/site-packages/opensoundscape/audio.py:1411: UserWarning: This seems to be an AudioMoth file, but parse_audiomoth_metadata() raised: 'comment'
  warnings.warn(


Epoch: 2 [batch 220/573, 38.39%] 
	Epoch Running Average Loss: 0.096
	Most Recent Batch Loss: 0.101
Epoch: 2 [batch 230/573, 40.14%] 
	Epoch Running Average Loss: 0.096
	Most Recent Batch Loss: 0.083
Epoch: 2 [batch 240/573, 41.88%] 
	Epoch Running Average Loss: 0.096
	Most Recent Batch Loss: 0.079
Epoch: 2 [batch 250/573, 43.63%] 
	Epoch Running Average Loss: 0.096
	Most Recent Batch Loss: 0.092
Epoch: 2 [batch 260/573, 45.38%] 
	Epoch Running Average Loss: 0.096
	Most Recent Batch Loss: 0.075
Epoch: 2 [batch 270/573, 47.12%] 
	Epoch Running Average Loss: 0.096
	Most Recent Batch Loss: 0.055
Epoch: 2 [batch 280/573, 48.87%] 
	Epoch Running Average Loss: 0.097
	Most Recent Batch Loss: 0.085
Epoch: 2 [batch 290/573, 50.61%] 
	Epoch Running Average Loss: 0.097
	Most Recent Batch Loss: 0.056


/Users/dawsonmccall/Documents/coding_projects/BioAcousticStreamEngine/.venv-training/lib/python3.11/site-packages/opensoundscape/audio.py:1411: UserWarning: This seems to be an AudioMoth file, but parse_audiomoth_metadata() raised: time data '01:00:00 14/08/202)' does not match format '%H:%M:%S %d/%m/%Y (%Z%z)'
  warnings.warn(


Epoch: 2 [batch 300/573, 52.36%] 
	Epoch Running Average Loss: 0.097
	Most Recent Batch Loss: 0.081
Epoch: 2 [batch 310/573, 54.10%] 
	Epoch Running Average Loss: 0.097
	Most Recent Batch Loss: 0.128
Epoch: 2 [batch 320/573, 55.85%] 
	Epoch Running Average Loss: 0.097
	Most Recent Batch Loss: 0.090
Epoch: 2 [batch 330/573, 57.59%] 
	Epoch Running Average Loss: 0.097
	Most Recent Batch Loss: 0.054
Epoch: 2 [batch 340/573, 59.34%] 
	Epoch Running Average Loss: 0.097
	Most Recent Batch Loss: 0.127
Epoch: 2 [batch 350/573, 61.08%] 
	Epoch Running Average Loss: 0.097
	Most Recent Batch Loss: 0.135
Epoch: 2 [batch 360/573, 62.83%] 
	Epoch Running Average Loss: 0.097
	Most Recent Batch Loss: 0.080
Epoch: 2 [batch 370/573, 64.57%] 
	Epoch Running Average Loss: 0.096
	Most Recent Batch Loss: 0.126
Epoch: 2 [batch 380/573, 66.32%] 
	Epoch Running Average Loss: 0.096
	Most Recent Batch Loss: 0.085
Epoch: 2 [batch 390/573, 68.06%] 
	Epoch Running Average Loss: 0.096
	Most Recent Batch Loss: 0.095


[src/libmpg123/layer3.c:INT123_do_layer3():1776] error: part2_3_length (3744) too large for available bit count (3736)


Epoch: 2 [batch 400/573, 69.81%] 
	Epoch Running Average Loss: 0.096
	Most Recent Batch Loss: 0.098


[src/libmpg123/layer3.c:INT123_do_layer3():1776] error: part2_3_length (1632) too large for available bit count (1560)


Epoch: 2 [batch 410/573, 71.55%] 
	Epoch Running Average Loss: 0.096
	Most Recent Batch Loss: 0.107
Epoch: 2 [batch 420/573, 73.30%] 
	Epoch Running Average Loss: 0.096
	Most Recent Batch Loss: 0.081
Epoch: 2 [batch 430/573, 75.04%] 
	Epoch Running Average Loss: 0.095
	Most Recent Batch Loss: 0.086
Epoch: 2 [batch 440/573, 76.79%] 
	Epoch Running Average Loss: 0.095
	Most Recent Batch Loss: 0.116
Epoch: 2 [batch 450/573, 78.53%] 
	Epoch Running Average Loss: 0.095
	Most Recent Batch Loss: 0.063
Epoch: 2 [batch 460/573, 80.28%] 
	Epoch Running Average Loss: 0.095
	Most Recent Batch Loss: 0.063
Epoch: 2 [batch 470/573, 82.02%] 
	Epoch Running Average Loss: 0.095
	Most Recent Batch Loss: 0.110
Epoch: 2 [batch 480/573, 83.77%] 
	Epoch Running Average Loss: 0.095
	Most Recent Batch Loss: 0.123
Epoch: 2 [batch 490/573, 85.51%] 
	Epoch Running Average Loss: 0.095
	Most Recent Batch Loss: 0.116
Epoch: 2 [batch 500/573, 87.26%] 
	Epoch Running Average Loss: 0.095
	Most Recent Batch Loss: 0.087


/Users/dawsonmccall/.pyenv/versions/3.11.9/lib/python3.11/tempfile.py:934: ResourceWarning: Implicitly cleaning up <TemporaryDirectory '/var/folders/dc/5p9wmvyd0m3c1sylx0z9lwlw0000gn/T/tmp3rakcl25wandb'>
  _warnings.warn(warn_message, ResourceWarning)
/Users/dawsonmccall/.pyenv/versions/3.11.9/lib/python3.11/tempfile.py:934: ResourceWarning: Implicitly cleaning up <TemporaryDirectory '/var/folders/dc/5p9wmvyd0m3c1sylx0z9lwlw0000gn/T/tmp9wzpndt5wandb'>
  _warnings.warn(warn_message, ResourceWarning)
/Users/dawsonmccall/.pyenv/versions/3.11.9/lib/python3.11/tempfile.py:934: ResourceWarning: Implicitly cleaning up <TemporaryDirectory '/var/folders/dc/5p9wmvyd0m3c1sylx0z9lwlw0000gn/T/tmpmhqa5jc_wandb-artifacts'>
  _warnings.warn(warn_message, ResourceWarning)
/Users/dawsonmccall/.pyenv/versions/3.11.9/lib/python3.11/tempfile.py:934: ResourceWarning: Implicitly cleaning up <TemporaryDirectory '/var/folders/dc/5p9wmvyd0m3c1sylx0z9lwlw0000gn/T/tmpvkrjrympwandb-artifacts'>
  _warnings.warn(war

Metrics:
	MAP: 0.731

Validation.


  0%|          | 0/202 [00:00<?, ?it/s]

/Users/dawsonmccall/.pyenv/versions/3.11.9/lib/python3.11/tempfile.py:934: ResourceWarning: Implicitly cleaning up <TemporaryDirectory '/var/folders/dc/5p9wmvyd0m3c1sylx0z9lwlw0000gn/T/tmpckw7rek_wandb'>
  _warnings.warn(warn_message, ResourceWarning)
/Users/dawsonmccall/.pyenv/versions/3.11.9/lib/python3.11/tempfile.py:934: ResourceWarning: Implicitly cleaning up <TemporaryDirectory '/var/folders/dc/5p9wmvyd0m3c1sylx0z9lwlw0000gn/T/tmpvwylwhi6wandb-artifacts'>
  _warnings.warn(warn_message, ResourceWarning)
/Users/dawsonmccall/.pyenv/versions/3.11.9/lib/python3.11/tempfile.py:934: ResourceWarning: Implicitly cleaning up <TemporaryDirectory '/var/folders/dc/5p9wmvyd0m3c1sylx0z9lwlw0000gn/T/tmpbdaw8w98wandb-media'>
  _warnings.warn(warn_message, ResourceWarning)
/Users/dawsonmccall/.pyenv/versions/3.11.9/lib/python3.11/tempfile.py:934: ResourceWarning: Implicitly cleaning up <TemporaryDirectory '/var/folders/dc/5p9wmvyd0m3c1sylx0z9lwlw0000gn/T/tmpuv5390n4wandb'>
  _warnings.warn(warn_me

Metrics:
	MAP: 0.506

Training Epoch 3


  0%|          | 0/573 [00:00<?, ?it/s]

/Users/dawsonmccall/Documents/coding_projects/BioAcousticStreamEngine/.venv-training/lib/python3.11/site-packages/torch/utils/data/dataloader.py:1095: UserWarning: 'pin_memory' argument is set as true but not supported on MPS now, device pinned memory won't be used.
  super().__init__(loader)


Epoch: 3 [batch 0/573, 0.00%] 
	Epoch Running Average Loss: 0.130
	Most Recent Batch Loss: 0.130


/Users/dawsonmccall/Documents/coding_projects/BioAcousticStreamEngine/.venv-training/lib/python3.11/site-packages/opensoundscape/audio.py:1411: UserWarning: This seems to be an AudioMoth file, but parse_audiomoth_metadata() raised: 'comment'
  warnings.warn(


Epoch: 3 [batch 10/573, 1.75%] 
	Epoch Running Average Loss: 0.092
	Most Recent Batch Loss: 0.070
Epoch: 3 [batch 20/573, 3.49%] 
	Epoch Running Average Loss: 0.091
	Most Recent Batch Loss: 0.070
Epoch: 3 [batch 30/573, 5.24%] 
	Epoch Running Average Loss: 0.090
	Most Recent Batch Loss: 0.050
Epoch: 3 [batch 40/573, 6.98%] 
	Epoch Running Average Loss: 0.089
	Most Recent Batch Loss: 0.063
Epoch: 3 [batch 50/573, 8.73%] 
	Epoch Running Average Loss: 0.089
	Most Recent Batch Loss: 0.091
Epoch: 3 [batch 60/573, 10.47%] 
	Epoch Running Average Loss: 0.087
	Most Recent Batch Loss: 0.087
Epoch: 3 [batch 70/573, 12.22%] 
	Epoch Running Average Loss: 0.085
	Most Recent Batch Loss: 0.061
Epoch: 3 [batch 80/573, 13.96%] 
	Epoch Running Average Loss: 0.086
	Most Recent Batch Loss: 0.093
Epoch: 3 [batch 90/573, 15.71%] 
	Epoch Running Average Loss: 0.085
	Most Recent Batch Loss: 0.101


[src/libmpg123/layer3.c:INT123_do_layer3():1776] error: part2_3_length (3744) too large for available bit count (3736)


Epoch: 3 [batch 100/573, 17.45%] 
	Epoch Running Average Loss: 0.084
	Most Recent Batch Loss: 0.049
Epoch: 3 [batch 110/573, 19.20%] 
	Epoch Running Average Loss: 0.084
	Most Recent Batch Loss: 0.070
Epoch: 3 [batch 120/573, 20.94%] 
	Epoch Running Average Loss: 0.084
	Most Recent Batch Loss: 0.079
Epoch: 3 [batch 130/573, 22.69%] 
	Epoch Running Average Loss: 0.083
	Most Recent Batch Loss: 0.094
Epoch: 3 [batch 140/573, 24.43%] 
	Epoch Running Average Loss: 0.083
	Most Recent Batch Loss: 0.042
Epoch: 3 [batch 150/573, 26.18%] 
	Epoch Running Average Loss: 0.083
	Most Recent Batch Loss: 0.074
Epoch: 3 [batch 160/573, 27.92%] 
	Epoch Running Average Loss: 0.083
	Most Recent Batch Loss: 0.067


/Users/dawsonmccall/Documents/coding_projects/BioAcousticStreamEngine/.venv-training/lib/python3.11/site-packages/opensoundscape/audio.py:1411: UserWarning: This seems to be an AudioMoth file, but parse_audiomoth_metadata() raised: time data '01:00:00 14/08/202)' does not match format '%H:%M:%S %d/%m/%Y (%Z%z)'
  warnings.warn(


Epoch: 3 [batch 170/573, 29.67%] 
	Epoch Running Average Loss: 0.083
	Most Recent Batch Loss: 0.109
Epoch: 3 [batch 180/573, 31.41%] 
	Epoch Running Average Loss: 0.083
	Most Recent Batch Loss: 0.068
Epoch: 3 [batch 190/573, 33.16%] 
	Epoch Running Average Loss: 0.084
	Most Recent Batch Loss: 0.056
Epoch: 3 [batch 200/573, 34.90%] 
	Epoch Running Average Loss: 0.084
	Most Recent Batch Loss: 0.063
Epoch: 3 [batch 210/573, 36.65%] 
	Epoch Running Average Loss: 0.085
	Most Recent Batch Loss: 0.055
Epoch: 3 [batch 220/573, 38.39%] 
	Epoch Running Average Loss: 0.085
	Most Recent Batch Loss: 0.117
Epoch: 3 [batch 230/573, 40.14%] 
	Epoch Running Average Loss: 0.085
	Most Recent Batch Loss: 0.132
Epoch: 3 [batch 240/573, 41.88%] 
	Epoch Running Average Loss: 0.085
	Most Recent Batch Loss: 0.105
Epoch: 3 [batch 250/573, 43.63%] 
	Epoch Running Average Loss: 0.085
	Most Recent Batch Loss: 0.041
Epoch: 3 [batch 260/573, 45.38%] 
	Epoch Running Average Loss: 0.085
	Most Recent Batch Loss: 0.138


[src/libmpg123/layer3.c:INT123_do_layer3():1776] error: part2_3_length (1920) too large for available bit count (1912)


Epoch: 3 [batch 460/573, 80.28%] 
	Epoch Running Average Loss: 0.085
	Most Recent Batch Loss: 0.110
Epoch: 3 [batch 470/573, 82.02%] 
	Epoch Running Average Loss: 0.085
	Most Recent Batch Loss: 0.068
Epoch: 3 [batch 480/573, 83.77%] 
	Epoch Running Average Loss: 0.085
	Most Recent Batch Loss: 0.055
Epoch: 3 [batch 490/573, 85.51%] 
	Epoch Running Average Loss: 0.085
	Most Recent Batch Loss: 0.087
Epoch: 3 [batch 500/573, 87.26%] 
	Epoch Running Average Loss: 0.085
	Most Recent Batch Loss: 0.077
Epoch: 3 [batch 510/573, 89.01%] 
	Epoch Running Average Loss: 0.085
	Most Recent Batch Loss: 0.080
Epoch: 3 [batch 520/573, 90.75%] 
	Epoch Running Average Loss: 0.085
	Most Recent Batch Loss: 0.111
Epoch: 3 [batch 530/573, 92.50%] 
	Epoch Running Average Loss: 0.085
	Most Recent Batch Loss: 0.037
Epoch: 3 [batch 540/573, 94.24%] 
	Epoch Running Average Loss: 0.085
	Most Recent Batch Loss: 0.064


[src/libmpg123/layer3.c:INT123_do_layer3():1776] error: part2_3_length (1632) too large for available bit count (1560)


Epoch: 3 [batch 550/573, 95.99%] 
	Epoch Running Average Loss: 0.085
	Most Recent Batch Loss: 0.040
Epoch: 3 [batch 560/573, 97.73%] 
	Epoch Running Average Loss: 0.085
	Most Recent Batch Loss: 0.093
Epoch: 3 [batch 570/573, 99.48%] 
	Epoch Running Average Loss: 0.085
	Most Recent Batch Loss: 0.057


/Users/dawsonmccall/.pyenv/versions/3.11.9/lib/python3.11/tempfile.py:934: ResourceWarning: Implicitly cleaning up <TemporaryDirectory '/var/folders/dc/5p9wmvyd0m3c1sylx0z9lwlw0000gn/T/tmp0tzqmkbwwandb'>
  _warnings.warn(warn_message, ResourceWarning)
/Users/dawsonmccall/.pyenv/versions/3.11.9/lib/python3.11/tempfile.py:934: ResourceWarning: Implicitly cleaning up <TemporaryDirectory '/var/folders/dc/5p9wmvyd0m3c1sylx0z9lwlw0000gn/T/tmpzu_w_27cwandb-artifacts'>
  _warnings.warn(warn_message, ResourceWarning)
/Users/dawsonmccall/.pyenv/versions/3.11.9/lib/python3.11/tempfile.py:934: ResourceWarning: Implicitly cleaning up <TemporaryDirectory '/var/folders/dc/5p9wmvyd0m3c1sylx0z9lwlw0000gn/T/tmp143w4k9fwandb-media'>
  _warnings.warn(warn_message, ResourceWarning)
/Users/dawsonmccall/.pyenv/versions/3.11.9/lib/python3.11/tempfile.py:934: ResourceWarning: Implicitly cleaning up <TemporaryDirectory '/var/folders/dc/5p9wmvyd0m3c1sylx0z9lwlw0000gn/T/tmpvi_49z2pwandb-media'>
  _warnings.warn(w

Metrics:
	MAP: 0.785

Validation.


  0%|          | 0/202 [00:00<?, ?it/s]

/Users/dawsonmccall/.pyenv/versions/3.11.9/lib/python3.11/tempfile.py:934: ResourceWarning: Implicitly cleaning up <TemporaryDirectory '/var/folders/dc/5p9wmvyd0m3c1sylx0z9lwlw0000gn/T/tmpaeil9k2lwandb'>
  _warnings.warn(warn_message, ResourceWarning)
/Users/dawsonmccall/.pyenv/versions/3.11.9/lib/python3.11/tempfile.py:934: ResourceWarning: Implicitly cleaning up <TemporaryDirectory '/var/folders/dc/5p9wmvyd0m3c1sylx0z9lwlw0000gn/T/tmphlhkjz7zwandb-artifacts'>
  _warnings.warn(warn_message, ResourceWarning)
/Users/dawsonmccall/.pyenv/versions/3.11.9/lib/python3.11/tempfile.py:934: ResourceWarning: Implicitly cleaning up <TemporaryDirectory '/var/folders/dc/5p9wmvyd0m3c1sylx0z9lwlw0000gn/T/tmpl3_1smonwandb-media'>
  _warnings.warn(warn_message, ResourceWarning)
/Users/dawsonmccall/.pyenv/versions/3.11.9/lib/python3.11/tempfile.py:934: ResourceWarning: Implicitly cleaning up <TemporaryDirectory '/var/folders/dc/5p9wmvyd0m3c1sylx0z9lwlw0000gn/T/tmp6mlddhovwandb-media'>
  _warnings.warn(w

Metrics:
	MAP: 0.546

Training Epoch 4


  0%|          | 0/573 [00:00<?, ?it/s]

/Users/dawsonmccall/Documents/coding_projects/BioAcousticStreamEngine/.venv-training/lib/python3.11/site-packages/torch/utils/data/dataloader.py:1095: UserWarning: 'pin_memory' argument is set as true but not supported on MPS now, device pinned memory won't be used.
  super().__init__(loader)


Epoch: 4 [batch 0/573, 0.00%] 
	Epoch Running Average Loss: 0.082
	Most Recent Batch Loss: 0.082
Epoch: 4 [batch 10/573, 1.75%] 
	Epoch Running Average Loss: 0.083
	Most Recent Batch Loss: 0.091
Epoch: 4 [batch 20/573, 3.49%] 
	Epoch Running Average Loss: 0.082
	Most Recent Batch Loss: 0.099
Epoch: 4 [batch 30/573, 5.24%] 
	Epoch Running Average Loss: 0.085
	Most Recent Batch Loss: 0.068
Epoch: 4 [batch 40/573, 6.98%] 
	Epoch Running Average Loss: 0.083
	Most Recent Batch Loss: 0.065
Epoch: 4 [batch 50/573, 8.73%] 
	Epoch Running Average Loss: 0.084
	Most Recent Batch Loss: 0.073
Epoch: 4 [batch 60/573, 10.47%] 
	Epoch Running Average Loss: 0.084
	Most Recent Batch Loss: 0.087
Epoch: 4 [batch 70/573, 12.22%] 
	Epoch Running Average Loss: 0.083
	Most Recent Batch Loss: 0.038
Epoch: 4 [batch 80/573, 13.96%] 
	Epoch Running Average Loss: 0.082
	Most Recent Batch Loss: 0.071
Epoch: 4 [batch 90/573, 15.71%] 
	Epoch Running Average Loss: 0.082
	Most Recent Batch Loss: 0.085


[src/libmpg123/layer3.c:INT123_do_layer3():1776] error: part2_3_length (1920) too large for available bit count (1912)


Epoch: 4 [batch 100/573, 17.45%] 
	Epoch Running Average Loss: 0.082
	Most Recent Batch Loss: 0.112
Epoch: 4 [batch 110/573, 19.20%] 
	Epoch Running Average Loss: 0.082
	Most Recent Batch Loss: 0.100
Epoch: 4 [batch 120/573, 20.94%] 
	Epoch Running Average Loss: 0.082
	Most Recent Batch Loss: 0.053
Epoch: 4 [batch 130/573, 22.69%] 
	Epoch Running Average Loss: 0.081
	Most Recent Batch Loss: 0.064
Epoch: 4 [batch 140/573, 24.43%] 
	Epoch Running Average Loss: 0.082
	Most Recent Batch Loss: 0.120
Epoch: 4 [batch 150/573, 26.18%] 
	Epoch Running Average Loss: 0.082
	Most Recent Batch Loss: 0.110


/Users/dawsonmccall/Documents/coding_projects/BioAcousticStreamEngine/.venv-training/lib/python3.11/site-packages/opensoundscape/audio.py:1411: UserWarning: This seems to be an AudioMoth file, but parse_audiomoth_metadata() raised: time data '01:00:00 14/08/202)' does not match format '%H:%M:%S %d/%m/%Y (%Z%z)'
  warnings.warn(


Epoch: 4 [batch 160/573, 27.92%] 
	Epoch Running Average Loss: 0.081
	Most Recent Batch Loss: 0.075
Epoch: 4 [batch 170/573, 29.67%] 
	Epoch Running Average Loss: 0.081
	Most Recent Batch Loss: 0.066
Epoch: 4 [batch 180/573, 31.41%] 
	Epoch Running Average Loss: 0.080
	Most Recent Batch Loss: 0.074
Epoch: 4 [batch 190/573, 33.16%] 
	Epoch Running Average Loss: 0.081
	Most Recent Batch Loss: 0.080
Epoch: 4 [batch 200/573, 34.90%] 
	Epoch Running Average Loss: 0.080
	Most Recent Batch Loss: 0.069
Epoch: 4 [batch 210/573, 36.65%] 
	Epoch Running Average Loss: 0.080
	Most Recent Batch Loss: 0.094
Epoch: 4 [batch 220/573, 38.39%] 
	Epoch Running Average Loss: 0.080
	Most Recent Batch Loss: 0.074
Epoch: 4 [batch 230/573, 40.14%] 
	Epoch Running Average Loss: 0.080
	Most Recent Batch Loss: 0.056
Epoch: 4 [batch 240/573, 41.88%] 
	Epoch Running Average Loss: 0.080
	Most Recent Batch Loss: 0.113
Epoch: 4 [batch 250/573, 43.63%] 
	Epoch Running Average Loss: 0.080
	Most Recent Batch Loss: 0.132


[src/libmpg123/layer3.c:INT123_do_layer3():1776] error: part2_3_length (1632) too large for available bit count (1560)


Epoch: 4 [batch 370/573, 64.57%] 
	Epoch Running Average Loss: 0.080
	Most Recent Batch Loss: 0.071
Epoch: 4 [batch 380/573, 66.32%] 
	Epoch Running Average Loss: 0.080
	Most Recent Batch Loss: 0.097
Epoch: 4 [batch 390/573, 68.06%] 
	Epoch Running Average Loss: 0.080
	Most Recent Batch Loss: 0.106
Epoch: 4 [batch 400/573, 69.81%] 
	Epoch Running Average Loss: 0.080
	Most Recent Batch Loss: 0.059
Epoch: 4 [batch 410/573, 71.55%] 
	Epoch Running Average Loss: 0.080
	Most Recent Batch Loss: 0.065
Epoch: 4 [batch 420/573, 73.30%] 
	Epoch Running Average Loss: 0.080
	Most Recent Batch Loss: 0.062
Epoch: 4 [batch 430/573, 75.04%] 
	Epoch Running Average Loss: 0.080
	Most Recent Batch Loss: 0.039
Epoch: 4 [batch 440/573, 76.79%] 
	Epoch Running Average Loss: 0.079
	Most Recent Batch Loss: 0.081


[src/libmpg123/layer3.c:INT123_do_layer3():1776] error: part2_3_length (3744) too large for available bit count (3736)


Epoch: 4 [batch 450/573, 78.53%] 
	Epoch Running Average Loss: 0.079
	Most Recent Batch Loss: 0.068
Epoch: 4 [batch 460/573, 80.28%] 
	Epoch Running Average Loss: 0.079
	Most Recent Batch Loss: 0.073
Epoch: 4 [batch 470/573, 82.02%] 
	Epoch Running Average Loss: 0.079
	Most Recent Batch Loss: 0.073
Epoch: 4 [batch 480/573, 83.77%] 
	Epoch Running Average Loss: 0.079
	Most Recent Batch Loss: 0.095
Epoch: 4 [batch 490/573, 85.51%] 
	Epoch Running Average Loss: 0.079
	Most Recent Batch Loss: 0.080


/Users/dawsonmccall/Documents/coding_projects/BioAcousticStreamEngine/.venv-training/lib/python3.11/site-packages/opensoundscape/audio.py:1411: UserWarning: This seems to be an AudioMoth file, but parse_audiomoth_metadata() raised: 'comment'
  warnings.warn(


Epoch: 4 [batch 500/573, 87.26%] 
	Epoch Running Average Loss: 0.079
	Most Recent Batch Loss: 0.054
Epoch: 4 [batch 510/573, 89.01%] 
	Epoch Running Average Loss: 0.079
	Most Recent Batch Loss: 0.078
Epoch: 4 [batch 520/573, 90.75%] 
	Epoch Running Average Loss: 0.079
	Most Recent Batch Loss: 0.087
Epoch: 4 [batch 530/573, 92.50%] 
	Epoch Running Average Loss: 0.079
	Most Recent Batch Loss: 0.050
Epoch: 4 [batch 540/573, 94.24%] 
	Epoch Running Average Loss: 0.078
	Most Recent Batch Loss: 0.065
Epoch: 4 [batch 550/573, 95.99%] 
	Epoch Running Average Loss: 0.078
	Most Recent Batch Loss: 0.100
Epoch: 4 [batch 560/573, 97.73%] 
	Epoch Running Average Loss: 0.078
	Most Recent Batch Loss: 0.069
Epoch: 4 [batch 570/573, 99.48%] 
	Epoch Running Average Loss: 0.078
	Most Recent Batch Loss: 0.060


/Users/dawsonmccall/.pyenv/versions/3.11.9/lib/python3.11/tempfile.py:934: ResourceWarning: Implicitly cleaning up <TemporaryDirectory '/var/folders/dc/5p9wmvyd0m3c1sylx0z9lwlw0000gn/T/tmp1clpnmmxwandb'>
  _warnings.warn(warn_message, ResourceWarning)
/Users/dawsonmccall/.pyenv/versions/3.11.9/lib/python3.11/tempfile.py:934: ResourceWarning: Implicitly cleaning up <TemporaryDirectory '/var/folders/dc/5p9wmvyd0m3c1sylx0z9lwlw0000gn/T/tmpusp9jq_pwandb-artifacts'>
  _warnings.warn(warn_message, ResourceWarning)
/Users/dawsonmccall/.pyenv/versions/3.11.9/lib/python3.11/tempfile.py:934: ResourceWarning: Implicitly cleaning up <TemporaryDirectory '/var/folders/dc/5p9wmvyd0m3c1sylx0z9lwlw0000gn/T/tmp32v_fr16wandb'>
  _warnings.warn(warn_message, ResourceWarning)
/Users/dawsonmccall/.pyenv/versions/3.11.9/lib/python3.11/tempfile.py:934: ResourceWarning: Implicitly cleaning up <TemporaryDirectory '/var/folders/dc/5p9wmvyd0m3c1sylx0z9lwlw0000gn/T/tmpqyuyishawandb-media'>
  _warnings.warn(warn_me

Metrics:
	MAP: 0.799

Validation.


  0%|          | 0/202 [00:00<?, ?it/s]

/Users/dawsonmccall/.pyenv/versions/3.11.9/lib/python3.11/tempfile.py:934: ResourceWarning: Implicitly cleaning up <TemporaryDirectory '/var/folders/dc/5p9wmvyd0m3c1sylx0z9lwlw0000gn/T/tmpuz2ny0k7wandb'>
  _warnings.warn(warn_message, ResourceWarning)
/Users/dawsonmccall/.pyenv/versions/3.11.9/lib/python3.11/tempfile.py:934: ResourceWarning: Implicitly cleaning up <TemporaryDirectory '/var/folders/dc/5p9wmvyd0m3c1sylx0z9lwlw0000gn/T/tmpezdc1hqnwandb-artifacts'>
  _warnings.warn(warn_message, ResourceWarning)
/Users/dawsonmccall/.pyenv/versions/3.11.9/lib/python3.11/tempfile.py:934: ResourceWarning: Implicitly cleaning up <TemporaryDirectory '/var/folders/dc/5p9wmvyd0m3c1sylx0z9lwlw0000gn/T/tmpb1jwbqauwandb-media'>
  _warnings.warn(warn_message, ResourceWarning)
/Users/dawsonmccall/.pyenv/versions/3.11.9/lib/python3.11/tempfile.py:934: ResourceWarning: Implicitly cleaning up <TemporaryDirectory '/var/folders/dc/5p9wmvyd0m3c1sylx0z9lwlw0000gn/T/tmpyzwy36m4wandb-media'>
  _warnings.warn(w

Metrics:
	MAP: 0.561

Best Model Appears at Epoch 4 with Validation score 0.561.


In [6]:
# ── Evaluate on held-out test set (3,137 clips) ──────────────────────────────
from sklearn.metrics import classification_report
import numpy as np
from opensoundscape.ml.cnn import load_model

scores = model.predict(test)
y_true = test.values
y_pred = (scores.values > 0.3).astype(int)
print(classification_report(
    y_true, y_pred,
    target_names=list(train.columns),
    zero_division=0,
))

# ── Evaluate best model on held-out test set (3,137 clips) ──────────────────────────────
best_model = torch.load(MODELS_DIR / "orthoptera_checkpoints" / "best.model", weights_only=False)

scores = best_model.predict(test)
y_true = test.values
y_pred = (scores.values > 0.3).astype(int)
print(classification_report(
    y_true, y_pred,
    target_names=list(train.columns),
    zero_division=0,
))

  0%|          | 0/4631 [00:00<?, ?it/s]

                                         precision    recall  f1-score   support

          Chorthippus brunneus brunneus       0.89      0.20      0.32        41
Pseudochorthippus parallelus parallelus       0.55      0.43      0.48       295
                    Omocestus viridulus       0.69      0.38      0.48        64
                 Tettigonia viridissima       0.82      0.64      0.72       887
                    Roeseliana roeselii       0.51      0.61      0.55       693
              Pholidoptera griseoaptera       0.87      0.54      0.66       509
               Leptophyes punctatissima       0.79      0.05      0.09       238
                     Gryllus campestris       0.96      0.84      0.89       861

                              micro avg       0.75      0.60      0.67      3588
                              macro avg       0.76      0.46      0.53      3588
                           weighted avg       0.77      0.60      0.65      3588
                          

  0%|          | 0/4631 [00:00<?, ?it/s]

                                         precision    recall  f1-score   support

          Chorthippus brunneus brunneus       0.89      0.20      0.32        41
Pseudochorthippus parallelus parallelus       0.55      0.43      0.48       295
                    Omocestus viridulus       0.69      0.38      0.48        64
                 Tettigonia viridissima       0.82      0.64      0.72       887
                    Roeseliana roeselii       0.51      0.61      0.55       693
              Pholidoptera griseoaptera       0.87      0.54      0.66       509
               Leptophyes punctatissima       0.79      0.05      0.09       238
                     Gryllus campestris       0.96      0.84      0.89       861

                              micro avg       0.75      0.60      0.67      3588
                              macro avg       0.76      0.46      0.53      3588
                           weighted avg       0.77      0.60      0.65      3588
                          

In [ ]:
# ── Threshold sweep — find F1-maximising cutoff per species ──────────────────
import numpy as np
import pandas as pd

thresholds = np.arange(0.1, 0.91, 0.05)
results = []
for t in thresholds:
    y_pred_t = (scores.values > t).astype(int)
    from sklearn.metrics import f1_score
    macro_f1 = f1_score(y_true, y_pred_t, average='macro', zero_division=0)
    micro_f1 = f1_score(y_true, y_pred_t, average='micro', zero_division=0)
    results.append({'threshold': round(t, 2), 'macro_f1': round(macro_f1, 3), 'micro_f1': round(micro_f1, 3)})

sweep_df = pd.DataFrame(results)
best = sweep_df.loc[sweep_df['macro_f1'].idxmax()]
print(f"Best macro-F1 threshold: {best['threshold']}  (macro={best['macro_f1']}, micro={best['micro_f1']})")
print()
print(sweep_df.to_string(index=False))


### Results — 5-epoch ResNet18 (CPU, May 2026)

| Species | Precision | Recall | F1 | Support |
|---|---|---|---|---|
| *Tettigonia viridissima* | 0.96 | 0.61 | **0.75** | 2,530 |
| *Chorthippus brunneus brunneus* | 0.81 | 0.23 | **0.36** | 73 |
| *Pholidoptera griseoaptera* | 0.76 | 0.12 | **0.21** | 1,393 |
| *Leptophyes punctatissima* | 0.85 | 0.08 | **0.15** | 359 |
| *Gryllus campestris* | 0.94 | 0.06 | **0.12** | 929 |
| *Roeseliana roeselii* | 0.59 | 0.06 | **0.10** | 741 |
| *Pseudochorthippus parallelus* | 0.77 | 0.03 | **0.05** | 358 |
| *Omocestus viridulus* | 1.00 | 0.01 | **0.03** | 76 |

**Macro F1: 0.22 — precision is high (low false-positive rate) but recall is low.**

Pattern: the model is conservative after only 5 epochs — it fires confidently when it fires,
but misses most detections. *Tettigonia* performs well (most training data: 6,270 clips).

**To improve:**
- More epochs (20–30) — the main lever
- Lower `min_confidence` threshold in `config/settings.yaml` (try 0.3)
- GPU training would make 30 epochs feasible in <1 hour

**Threshold sweep verdict:** macro-F1 is flat across all thresholds (0.215–0.225). Lowering the threshold does not help — the model's score distribution is too compressed after only 5 epochs. More training epochs is the only meaningful lever.


### Results — 5-epoch ResNet18 (GPU, June 23rd 2026)
### Alteration: fixes for Multi-hot Encoding and proper start and end times

| Species | Precision | Recall | F1 | Support |
|---|---|---|---|---|
| *Tettigonia viridissima* | 0.83 | 0.62 | **0.71** | 887 |
| *Chorthippus brunneus brunneus* | 0.93 | 0.34 | **0.50** | 41 |
| *Pholidoptera griseoaptera* | 0.99 | 0.41 | **0.58** | 509 |
| *Leptophyes punctatissima* | 0.61 | 0.15 | **0.24** | 238 |
| *Gryllus campestris* | 0.98 | 0.91 | **0.94** | 861 |
| *Roeseliana roeselii* | 0.51 | 0.50 | **0.50** | 693 |
| *Pseudochorthippus parallelus* | 0.89 | 0.37 | **0.53** | 295 |
| *Omocestus viridulus* | 0.97 | 0.55 | **0.70** | 64 |

**Macro F1: 0.59.**

### Results — 5-epoch ResNet18 (GPU, June 24th 2026)
### Alterations: splitting InsectSet459 into 4 second clips like EcoSoundSet

| Species | Precision | Recall | F1 | Support |
|---|---|---|---|---|
| *Tettigonia viridissima* | 0.85 | 0.76 | **0.80** | 887 |
| *Chorthippus brunneus brunneus* | 0.64 | 0.44 | **0.52** | 41 |
| *Pholidoptera griseoaptera* | 0.84 | 0.49 | **0.62** | 509 |
| *Leptophyes punctatissima* | 0.50 | 0.03 | **0.06** | 238 |
| *Gryllus campestris* | 0.98 | 0.92 | **0.95** | 861 |
| *Roeseliana roeselii* | 0.73 | 0.24 | **0.36** | 693 |
| *Pseudochorthippus parallelus* | 0.75 | 0.31 | **0.44** | 295 |
| *Omocestus viridulus* | 0.95 | 0.66 | **0.78** | 64 |

**Macro F1: 0.56**

### Results — 5-epoch ResNet18 (GPU, June 25th 2026)
### Alterations: taking the 4s window with the greatest energy above 2 kHz frequency for each InsectSet459 clip.

Many InsectSet clips are poor quality or mostly silence, so instead of using the first 4s or splitting the whole clip into 4s segments we target higher frequencies (> 2000 Hz) and take the 4s segment with the most noise in that range. As the metrics show, this change slightly improved overall F1 Macro:

| Species | Precision | Recall | F1 | Support |
|---|---|---|---|---|
| *Tettigonia viridissima* | 0.85 | 0.76 | **0.80** | 887 |
| *Chorthippus brunneus brunneus* | 0.64 | 0.44 | **0.52** | 41 |
| *Pholidoptera griseoaptera* | 0.84 | 0.49 | **0.62** | 509 |
| *Leptophyes punctatissima* | 0.50 | 0.03 | **0.06** | 238 |
| *Gryllus campestris* | 0.98 | 0.92 | **0.95** | 861 |
| *Roeseliana roeselii* | 0.73 | 0.24 | **0.36** | 693 |
| *Pseudochorthippus parallelus* | 0.75 | 0.31 | **0.44** | 295 |
| *Omocestus viridulus* | 0.95 | 0.66 | **0.78** | 64 |
| **Micro avg** | 0.85 | 0.54 | **0.66** | 3588 |
| **Macro avg** | 0.83 | 0.49 | **0.60** | 3588 |
| **Weighted avg** | 0.83 | 0.54 | **0.63** | 3588 |
| **Samples avg** | 0.60 | 0.57 | **0.58** | 3588 |

### Results — 5-epoch ResNet34 (GPU, June 29th 2026)
### Alterations: first training attempt with the ResNet34 architecture.

| Species | Precision | Recall | F1 | Support |
|---|---|---|---|---|
| *Tettigonia viridissima* | 0.83 | 0.50 | **0.62** | 887 |
| *Chorthippus brunneus brunneus* | 1.00 | 0.54 | **0.70** | 41 |
| *Pholidoptera griseoaptera* | 0.97 | 0.50 | **0.65** | 509 |
| *Leptophyes punctatissima* | 0.30 | 0.08 | **0.13** | 238 |
| *Gryllus campestris* | 0.97 | 0.94 | **0.96** | 861 |
| *Roeseliana roeselii* | 0.54 | 0.56 | **0.55** | 693 |
| *Pseudochorthippus parallelus* | 0.82 | 0.37 | **0.51** | 295 |
| *Omocestus viridulus* | 0.94 | 0.70 | **0.80** | 64 |
| **Micro avg** | 0.80 | 0.58 | **0.68** | 3588 |
| **Macro avg** | 0.80 | 0.52 | **0.62** | 3588 |
| **Weighted avg** | 0.80 | 0.58 | **0.66** | 3588 |
| **Samples avg** | 0.64 | 0.62 | **0.63** | 3588 |

### Results — 5-epoch ResNet34 (GPU, June 30th 2026)
### Alterations: Include negative clips (clips with no Orthoptera annotations whatsoever) to help model ignore anthropogenic and non-target biotic sounds in target Orthoptera clips.

| Species | Precision | Recall | F1 | Support |
|---|---|---|---|---|
| *Tettigonia viridissima* | 0.82 | 0.64 | **0.72** | 887 |
| *Chorthippus brunneus brunneus* | 0.89 | 0.20 | **0.32** | 41 |
| *Pholidoptera griseoaptera* | 0.87 | 0.54 | **0.66** | 509 |
| *Leptophyes punctatissima* | 0.79 | 0.05 | **0.09** | 238 |
| *Gryllus campestris* | 0.96 | 0.84 | **0.89** | 861 |
| *Roeseliana roeselii* | 0.51 | 0.61 | **0.55** | 693 |
| *Pseudochorthippus parallelus* | 0.55 | 0.43 | **0.48** | 295 |
| *Omocestus viridulus* | 0.69 | 0.38 | **0.48** | 64 |
| **Micro avg** | 0.75 | 0.60 | **0.67** | 3588 |
| **Macro avg** | 0.76 | 0.46 | **0.53** | 3588 |
| **Weighted avg** | 0.77 | 0.60 | **0.65** | 3588 |
| **Samples avg** | 0.45 | 0.43 | **0.43** | 3588 |

In [ ]:
# ── Save final model ──────────────────────────────────────────────────────────
# OpenSoundscape saves as a .model file (a torch pickle).
# This is the file you point insect.py at in BASE.

model_path = MODELS_DIR / "orthoptera_uk.model"
model.save(model_path)
print(f"Model saved: {model_path}")
print()
print("Next step: set 'model_path' in config/settings.yaml under 'insect:'")
print("and activate 'insect' in classifiers.active")